components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/GraphBasedSequenceHandler.java

```diff
@@ -144,24 +144,66 @@ private void handleEndOfSequence(HttpServletRequest request, HttpServletResponse
 
     /**
      * Process FailNode.
-     * @param request
-     * @param response
-     * @param context
-     * @param sequenceConfig
-     * @param node
+     * @param request HTTP Servlet request
+     * @param response HTTP Servlet Response
+     * @param context Authentication Context
+     * @param sequenceConfig Sequence Config
+     * @param node Fail Node
      * @throws FrameworkException
      */
     private void handleAuthFail(HttpServletRequest request, HttpServletResponse response, AuthenticationContext context,
-            SequenceConfig sequenceConfig, AuthGraphNode node) throws FrameworkException {
+            SequenceConfig sequenceConfig, FailNode node) throws FrameworkException {
 
         if (log.isDebugEnabled()) {
             log.debug("Found a Fail Node in conditional authentication");
         }
+
+        AuthGraphNode stepNode;
+
+        // Check if the retrying is enabled.
+        if (isRetryEnabled(context)) {
+            if (node != null && node.getPrevious() != null) {
+                AuthGraphNode previous = node.getPrevious();
+                if (previous instanceof DynamicDecisionNode && ((DynamicDecisionNode) previous).getPrevious() != null) {
+                    stepNode = ((DynamicDecisionNode) previous).getPrevious();
+                    if (stepNode instanceof StepConfigGraphNode) {
+                        resetGraphForRetrying(request, context, (StepConfigGraphNode)stepNode);
+                        return;
+                    }
+                } else if (previous instanceof StepConfigGraphNode) {
+                    resetGraphForRetrying(request, context, (StepConfigGraphNode)previous);
+                    return;
+                }
+            }
+        }
         context.setRequestAuthenticated(false);
         context.getSequenceConfig().setCompleted(true);
         //TODO:Need to consider sending error message to provided error page Uri instead of oauth redirect Uri
     }
 
+    private boolean isRetryEnabled(AuthenticationContext context) {
+
+        if (context.getProperty(FrameworkConstants.JSAttributes.JS_RETRY_STEP) == null) {
+            // If the retry property is not set in the authentication context, it is assumed to be true.
+            return true;
```

Comment:
what would happen if the authenticators don't support retrying? Will it force them to retry?

Reply:
Created https://github.com/wso2/product-is/issues/3161 to track the issue.